### PR TITLE
Revamp the soap client connector

### PR DESCRIPTION
**Pull request by Bhashinee**
## Purpose
- Revamp the SOAP client connector and fix identified issues with the current implementation.
- Add passwordDigest WS-Security to the SOAP client.
- Fix : #40, #39, #37, #36, #38

## Goals
- Address drawbacks and identified issues in the SOAP client connector.
- Revamp the API into a more user-friendly way. 
- Introduce Password digest WS-Security.

## Approach
- Instead of having 1 client for both SOAP 1.1 and SOAP 1.2, have 2 different clients as 'soap11Client' and 'soap12Client'. So that we can mandate relevant parameters in each client.
- Group the related fields in the current SoapRequest record and have them in separate records under the Options record.

## Samples
```
import ballerina/io;
import wso2/soap;

public function main() {
    soap:Soap11Client soapClient = new("http://localhost:9000");

    xml body = xml `<m0:getQuote xmlns:m0="http://services.samples">
                        <m0:request>
                            <m0:symbol>WSO2</m0:symbol>
                        </m0:request>
                    </m0:getQuote>`;

    soap:UsernameToken usernameToken = {
        username: "admin",
        password: "admin",
        passwordType: "PasswordDigest"
    };

    soap:Options options = {
    	usernameToken: usernameToken
    };

    var response = soapClient->sendReceive("/services/SimpleStockQuoteService", "urn:mediate", body, options = options);
    if (response is soap:SoapResponse) {
        io:println(response);
    } else {
        io:println(response.detail().message);
    }
}
```
Above is an example of SOAP 1.1 client, connecting to a SOAP service which is protected by usernameToken passwordDigest WS-Security. Similarly we can create a SOAP1.2 client.

## Related issues
https://github.com/ballerina-platform/ballerina-lang/issues/14587

### CODE REVIEW

File: soap/soap11_connector.bal [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// under the License.
+
+import ballerina/http;
+
+# SOAP 1.1 client connector.
+#
+# + soap11Client - HTTP client endpoint for SOAP 1.1 client
+public type Soap11Connector client object {
+
+    public http:Client soap11Client;
```
Comment (wggihan) [inline soap/soap11_connector.bal]: Why this client is public?

File: soap/soap11_connector.bal [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    remote function sendReceive(string path, string soapAction, xml body, Options? options = ())
+            returns SoapResponse|error;
+
+    remote function sendRobust(string path, string soapAction, xml body, Options? options = ()) returns error?;
+
+    remote function fireAndForget(string path, string soapAction, xml body, Options? options = ());
+};
+
+remote function Soap11Connector.sendReceive(string path, string soapAction, xml body, Options? options = ())
```
Comment (wggihan) [inline soap/soap11_connector.bal]: Shall we move this code to the client object itself? @hasithaa what is the recommended approach?

File: soap/soap11_connector.bal [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+# SOAP 1.1 client connector.
+#
+# + soap11Client - HTTP client endpoint for SOAP 1.1 client
+public type Soap11Connector client object {
```
Comment (wggihan) [inline soap/soap11_connector.bal]: Let's remove this connector and just use client only.

File: soap/soap11_connector.bal (hunk 2 above)
Comment (Bhashinee) [reply soap/soap11_connector.bal]: Moved to the client object.

File: soap/soap11_endpoint.bal [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    # Sends SOAP 1.1 request and expects a response.
+    #
+    # + path - Resource path
+    # + soapAction - SOAP action
+    # + body - SOAP payload
+    # + options - SOAP options. Ex: Headers, Ws-addressing parameters, usernameToken parameters
+    # + return - If a success, returns the response object, else returns `SoapError` object
+    public remote function sendReceive(string path, string soapAction, xml body, Options? options = ())
+            returns SoapResponse|error {
+        http:Client httpClient = self.soap11Client;
```
Comment (wggihan) [inline soap/soap11_endpoint.bal]: Why we need this assignment? Can we directly pass the soap11Client?

File: soap/soap11_endpoint.bal [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    # Fire and forget requests. Sends the request without the possibility of any response from the
+    # service(even an error).
+    #
+    # + path - Resource path
+    # + options - SOAP options. Ex: Headers, Ws-addressing parameters, usernameToken parameters
+    public remote function fireAndForget(string path, string soapAction, xml body, Options? options = ()) {
+        http:Client httpClient = self.soap11Client;
+        return fireAndForget(path, soapAction = soapAction, body, options = options, httpClient, SOAP11);
```
Comment (wggihan) [inline soap/soap11_endpoint.bal]: Why we need a return statement here? We haven't specified any returns in the API signature.

File: soap/soap_types.bal [hunk 6]
```diff
@@ -21,10 +21,28 @@ public const SOAP12 = "SOAP12";
 
 # Represents SOAP request.
 #
-# + soapVersion - The version of SOAP
-# + soapAction - The value of SOAP Action to send to the endpoint
-# + payload - The xml of SOAP payload for the SOAP envelop to send to the endpoint
 # + headers - The array of SOAP headers for the SOAP envelop to send to the endpoint
+# + wsAddressing - Soap WS-Addressing related options.
+# + usernameToken - Soap WS-Username token related options.
+public type Options record {
```
Comment (wggihan) [inline soap/soap_types.bal]: Don't we need to seal this record?

File: soap/soap_utils.bal [hunk 7]
```diff
@@ -112,63 +115,83 @@ function getWSAddressingHeaders(SoapRequest request) returns xml {
 
 # Provides the WS secure username token headers.
 #
-# + request - The request to be sent
+# + options - Soap options to be sent
 # + return - XML with the WS secure username token headers
-function getWSSecreUsernameTokenHeaders(SoapRequest request) returns xml {
+function getWSSecreUsernameTokenHeaders(Options options) returns xml {
```
Comment (chamil321) [inline soap/soap_utils.bal]:
```suggestion
function getWSSecureUsernameTokenHeaders(Options options) returns xml {
```

File: soap/soap_utils.bal [hunk 8]
```diff
@@ -112,63 +115,83 @@ function getWSAddressingHeaders(SoapRequest request) returns xml {
...
+            } else {
+                headerElement = headerElement + getWSAddressingHeaders(options);
+            }
         }
-
-        if (request["username"] is string) {
-            headerElement += getWSSecreUsernameTokenHeaders(request);
+        if (options["usernameToken"]["username"] is string) {
+            if (headerElement is ()) {
+                headerElement = getWSSecreUsernameTokenHeaders(options);
```
Comment (chamil321) [inline soap/soap_utils.bal]:
```suggestion
                headerElement = getWSSecureUsernameTokenHeaders(options);
```

File: soap/soap_utils.bal [hunk 9]
```diff
@@ -112,63 +115,83 @@ function getWSAddressingHeaders(SoapRequest request) returns xml {
...
+            } else {
+                headerElement = headerElement + getWSSecreUsernameTokenHeaders(options);
```
Comment (chamil321) [inline soap/soap_utils.bal]:
```suggestion
                headerElement = headerElement + getWSSecureUsernameTokenHeaders(options);
```

File: soap/soap_utils.bal [hunk 10]
```diff
@@ -241,3 +277,46 @@ function createSOAPResponse(http:Response response, SoapVersion soapVersion) ret
...
 }
+
+# Creates the password used in password digest usernameToken WS-Security.
+#
+# + nonce - The nonce string
+# + password - The password in plain text
+# + createdTime - The created timestamp
+# + return - The digest password in string format
+function createDigestPassword(string nonce, string password, string createdTime) returns string {
+    string concatenatedDigest = nonce+createdTime+password;
```
Comment (chamil321) [inline soap/soap_utils.bal]:
```suggestion
    string concatenatedDigest = nonce + createdTime + password;
```

File: soap/soap11_endpoint.bal [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+# Object for SOAP 1.1 client endpoint.
+#
+# + soap11Client - Http client created to send SOAP 1.1 requests.
+public type Soap11Client client object {
+
+    private http:Client soap11Client;
+
+    public function __init(string url, SoapConfiguration? soapConfig = ()) {
+        self.soap11Client = new(url, config = soapConfig.clientConfig);
```
Comment (chamil321) [inline soap/soap11_endpoint.bal]: Shouldn't we type check the soapConfig here? What if nill was passed?

Comment (Bhashinee) [reply soap/soap11_endpoint.bal]: This is type Checked in HTTP client initialization.

**Comment by praneesha** [comment]
Issue https://github.com/wso2-ballerina/module-soap/issues/44 created for the doc review of this.

**Pull request merged**